### ISSUE TITLE
fastai parity

### ISSUE THREAD

**Issue by lorenzoh** [FluxML/ML-Coordination-Tracker#16]
This issue tracks the progress on fastai parity.

The goal is to
- catalog all features of fastai
- categorize them
- evaluate the progress toward having that feature in the FastAI.jl ecosystem
- estimate the remaining effort
- prioritize them

and thereby make it easier to contribute to FastAI.jl.

The first step is to catalog all the features fastai has. The relevant resources for this are

- [the documentation](https://docs.fast.ai)
- [the book chapters](https://github.com/fastai/fastbook)
- [the video lectures](https://course19.fast.ai/part2),especially part 2

Help is much appreciated!

**Comment by lorenzoh** [comment]
Updated the list with pointers to existing implementations

**Comment by ToucheSir** [comment]
I vote for reviving MLMetrics instead of rolling them into FluxTraining. I'm sure there would be many base Flux, Knet and other library users who could make use of them. Currently everybody has a incomplete subset of metrics with incompatible APIs (e.g. [Avalon.jl](https://github.com/dfdx/Lilith.jl/blob/bbd8e8147c2d71f175ced32855ca19f3607f0baa/src/metrics.jl)). I would hope that we could get rid of those for a "NNlib of metrics" instead.

**Comment by DoktorMike** [comment]
> I vote for reviving MLMetrics instead of rolling them into FluxTraining. I'm sure there would be many base Flux, Knet and other library users who could make use of them. Currently everybody has a incomplete subset of metrics with incompatible APIs (e.g. [Avalon.jl](https://github.com/dfdx/Lilith.jl/blob/bbd8e8147c2d71f175ced32855ca19f3607f0baa/src/metrics.jl)). I would hope that we could get rid of those for a "NNlib of metrics" instead.

Since most of these metrics are really distances in one form or another wouldn't it be natural to use the existing implementation in [Distances.jl](https://github.com/JuliaStats/Distances.jl) ? But maybe I'm missing something.

**Comment by darsnack** [comment]
Yeah I think implementing the distance part of it in Distances.jl makes a lot of sense.

**Comment by ToucheSir** [comment]
I think there's still a need for a metrics package because most of the classification metrics don't make sense in Distances.jl. There's also the case of domain-specific metrics like Dice score and BLEU.

**Comment by CarloLucibello** [comment]
there should be some distinction between losses and metrics?

**Comment by ToucheSir** [comment]
At the very least, the metrics package/namespace could reexport losses that are also metrics.

**Comment by darsnack** [comment]
Yeah I think the hierarchy would be Distances -> Metrics -> Losses. There will be losses that are not metrics (i.e. defined completely in a loss package), and losses that just reexport a metric. Similarly there will be metrics that are completely defined the metrics package, but many will reexport or build upon standard distances.

To that end, I agree that we should make use of Distances.jl as much as possible. And if there is a metric that generalizes to a distance, then we can submit a PR to Distances.jl.

**Comment by lorenzoh** [comment]
I agree with @darsnack! Every loss can be a metric (not in a strict mathematical sense but as a measure of model performance), but not the reverse.

**Comment by lorenzoh** [comment]
Also, if MLMetrics.jl is revived, FluxTraining.jl should depend on it.

**Comment by lorenzoh** [comment]
Updated the list

**Comment by darsnack** [comment]
@lorenzoh do you want to transfer stuff from here to FastAI.jl issues?
